rfcs/20190729-keras-preprocessing-redesign.md

```diff
@@ -0,0 +1,351 @@
+    """
+
+def from_dataframe(
+    dataframe,
+    directory=None,
+    data_column='filename',
+    target_column='class',
+    target_mode='categorical',
+    weight_column=None,
+    color_mode='rgb',
+    batch_size=32,
+    image_size=(255, 255),
+    shuffle=True,
+    seed=None,
+    validation_split=None,
+    subset=None):
+    """Generates a Dataset from a Pandas dataframe.
+
+    # Arguments:
+        dataframe: Pandas dataframe instance.
+        directory: The directory that image paths refer to.
+        data_column: Name of column with the paths for the input images.
+        target_column: Name of column with the class information.
+        target_mode:
+            - 'categorical' means that the inferred labels are
+                encoded as a categorical vector (e.g. for categorical_crossentropy).
+            - 'binary' means that the inferred labels (there can be only 2)
+                are encoded as binary scalars (e.g. for binary_crossentropy).
+        weight_column: Name of column with sample weight information.
+        color_mode: One of "grayscale", "rgb", "rgba". Default: "rgb".
+            Whether the images will be converted to
+            have 1, 3, or 4 channels.
+        batch_size: Size of the batches of data (default: 32).
+        image_size: Size to resize images to after they are read from disk.
+          Since the pipeline processes batches of images that must all have the same size,
+          this must be provided.
+        shuffle: Whether to shuffle the data (default: True)
+            If set to False, sorts the data in alphanumeric order.
+        seed: Optional random seed for shuffling and transformations.
+        validation_split: Optional float between 0 and 1,
+            fraction of data to reserve for validation.
+        subset: One of "training" or "validation". Only used if `validation_split` is set.
+    """
+
+def preview(self, data, save_to_directory=None, save_prefix=None, save_format='png'):
+    """Enables users to preview the image augmentation configuration.
+
+    # Arguments
+        data: Image data. Could be strings (a list of image paths), a list of PIL image instances,
+            a list of arrays, or a list of eager tensors.
+        save_to_directory: Directory to save transformed images. Mandatory if not in a notebook.
+            If in a notebook and this is not specified, images are displayed in-line.
+        save_prefix: String, filename prefix for saved images.
+        save_format: String, extension for saved images.
+    """
+```
+
+**Note:** `from_arrays` is not included since it is possible to transform Numpy data simply by calling the `ImagePipeline` object (like a layer).
+
+
+### Layers
+
+The new data augmentation layers will inherit `keras.layers.Layer` and work in a similar way.
+
+```python
+Resizing(height, width)  # Resize while distorting aspect ratio
+CenterCrop(height, width)  # Resize without distorting aspect ratio
```

Comment:
Is there value in a RandomCrop? Or just Crop, with center v random parameterized? IIRC, random cropping is part of some imagenet pipelines.

Reply:
I think adding `RandomCrop` is a good idea. It is technically equivalent to a combination of `RandomTranslation` and `CenterCrop`, but it is a useful shortcut.